### PR TITLE
feat(Core/Creature): Dual id Spawning WIP

### PR DESCRIPTION
**Pull request by malcrom**
## Changes Proposed:
-  Add the ability to assign two entries to a guid for random spawning of either creature.
- This is up atm for my code to be reviewed and possibly cleaned up. I will add SQL to change creature table and test on Skeletal Horrors & Fiends from Duskwood later today.

## Issues Addressed:
- Closes 

## SOURCE:
Anyone who sniffs can confirm some spawn points can be one of two entries.

## Tests Performed:
- Been testing in game. I can't find any more issues atm so I'm putting this up for testing.

## How to Test the Changes:
1. Will update with SQL later today

### CODE REVIEW

**Review by acidmanifesto — APPROVED** [review]

**Comment by acidmanifesto** [comment]
For those who are just now following and read this PR and is curious what it's purpose is, The sniff teams observation has confirmed and shown that when select creatures respawn, at times they will respawn as a different creature entirely and not the same creature.

File: src/server/game/Globals/ObjectMgr.cpp [hunk 1]
```diff
@@ -2016,38 +2016,52 @@ void ObjectMgr::LoadCreatures()
...
+            LOG_ERROR("sql.sql", "Table `creature` has creature (SpawnId: %u) with non existing creature entry %u in creature_id1 field, skipped.", spawnId, id);
+            continue;
+        }
+        CreatureTemplate const* cInfo2 = GetCreatureTemplate(id2);
+        if (!cInfo2 && id2)
+        {
+            LOG_ERROR("sql.sql", "Table `creature` has creature (SpawnId: %u) with non existing creature entry %u in creature_id2 field, skipped.", spawnId, id2);
+            continue;
+        }
+        if (chance <= 0 || chance > 100)
```
Comment (Kitzunu) [inline src/server/game/Globals/ObjectMgr.cpp:2035]:
```suggestion
        if (!chance || chance > 100)
```

Comment (Kitzunu) [reply src/server/game/Globals/ObjectMgr.cpp:2035]: Unsigned integers cannot go below 0 so we don't need to check for that

File: src/server/game/Globals/ObjectMgr.cpp [hunk 2]
```diff
@@ -2016,38 +2016,52 @@ void ObjectMgr::LoadCreatures()
...
+        {
+            LOG_ERROR("sql.sql", "Table `creature` chance_id1 (Value: %u) must be greater than 0 and less than or equal to 100, skipped.", chance);
             continue;
         }
```
Comment (Kitzunu) [inline src/server/game/Globals/ObjectMgr.cpp:2040]:
```suggestion
        if (chance == 100 && id2)
        {
            LOG_ERROR("sql.sql", "Table `creature` has spawnid = %u chance = 100 even though creature_id2 has an entry. changed to 50.", chance);
            chance = 50;
```

Comment (Kitzunu) [reply src/server/game/Globals/ObjectMgr.cpp:2040]: chance can never be 100 if we have a creature id2 in the database. Just wouldn't work, so better if we just throw an error and change the chance to 50.

I don't think we need to skip load though

File: data/sql/updates/pending_db_world/rev_1641837958335217980.sql [hunk 3]
```diff
@@ -0,0 +1,9 @@
+INSERT INTO `version_db_world` (`sql_rev`) VALUES ('1641837958335217980');
+
+ALTER TABLE `creature`
+    CHANGE COLUMN `id` `creature_id1` MEDIUMINT UNSIGNED NOT NULL DEFAULT 0 COMMENT 'Creature Identifier' AFTER `guid`;
+ALTER TABLE `creature` 
+    ADD COLUMN `creature_id2` MEDIUMINT UNSIGNED NOT NULL DEFAULT 0 COMMENT 'Creature Identifier' AFTER `creature_id1`;
+ALTER TABLE `creature` 
+    ADD COLUMN `chance_id1` FLOAT UNSIGNED NOT NULL DEFAULT 100 COMMENT 'Chance id1 spawns' AFTER `creature_id2`;
+
```
Comment (Kitzunu) [inline data/sql/updates/pending_db_world/rev_1641837958335217980.sql:9]:
```suggestion
    ADD COLUMN `creature_id2` MEDIUMINT UNSIGNED NOT NULL DEFAULT 0 COMMENT 'Creature Identifier' AFTER `creature_id1`,
    ADD COLUMN `chance_id1` FLOAT UNSIGNED NOT NULL DEFAULT 100 COMMENT 'Chance id1 spawns' AFTER `creature_id2`;

```

**Comment by malcrom** [comment]
Test code 

```
-- Skeletal Horrors and Skeletal Fiends, Duskwood
UPDATE `creature` SET `Creature_id1`=202,`Creature_id2`=531, `chance_id1`=50 WHERE `guid` IN 
(4927,4941,4943,4944,4945,4946,4969,4972,4973,4974,4993,5038,5039,5040,5041,5042,5043,5139,5140,5160,5163,5168,5176,5178,5942,5958,5959,5961);
```

File: src/server/game/Globals/ObjectMgr.cpp (hunk 2 above)
Comment (malcrom) [reply src/server/game/Globals/ObjectMgr.cpp:2040]: Agreed

File: src/server/game/Entities/Creature/Creature.cpp [hunk 4]
```diff
@@ -1900,26 +1905,34 @@ void Creature::Respawn(bool force)
...
 
         setDeathState(JUST_RESPAWNED);
-        // MDic - Acidmanifesto
-        // If creature has genders it will consider gender changing on respawn.
-        if (sObjectMgr->GetCreatureTemplate(m_originalEntry))
-        {
-            InitEntry(m_originalEntry);
-        }
 
+        // MDic - Acidmanifesto
```
Comment (acidmanifesto) [inline src/server/game/Entities/Creature/Creature.cpp:1935]: Remove note since u reworked it.

**Comment by claudiodfc** [comment]
Great work Malcrom. I wonder if two creatures is the ideal? Vmangos has at least 4 and we need 3 to fix Devilsaur spawns in Ungoro for Tyrant, Ironhide and regular Devilsaur.

**Pull request merged**